sdk/digitaltwins/Azure.DigitalTwins.Core/src/Queries/QueryBuilders/Linq/DigitalTwinsQueryBuilder.cs

```diff
@@ -0,0 +1,406 @@
+                {
+                    throw new InvalidOperationException();
+                }
+            }
+
+            string text = string.Format(formatProvider, filter.Format, args);
+            return Where(text);
+        }
+
+        /// <summary>
+        /// Adds a WHERE clause to a query along with conditions defined with LINQ expressions.
+        /// </summary>
+        /// <param name="filter">Conditional argument relevant to a DigitalTwins query (e.g. comparisons, DigitalTwins Functions, etc.)</param>
+        /// <returns>Query that contains a WHERE clause and conditional arguments.</returns>
+        public DigitalTwinsQueryBuilder<T> Where(Expression<Func<T, bool>> filter)
+        {
+            string query = string.Empty;
+            Expression e = Evaluator.PartialEval(filter, CanEvaluate);
+            e = ExpressionNormalizer.Normalize(e, new Dictionary<Expression, Expression>());
+
+            LambdaExpression l = e as LambdaExpression;
+            Ensure(l != null);
+            Ensure(l.Parameters.Count == 1);
+
+            query = QueryFilterVisitor.Translate(l.Body);
+
+            return Where(query);
+
+            bool CanEvaluate(Expression e)
+            {
+                if (e is MethodCallExpression call && call.Method.DeclaringType == typeof(DigitalTwinsFunctions))
+                {
+                    return false;
+                }
+
+                return Evaluator.CanBeEvaluatedLocally(e);
+            }
+        }
+
+        private static void Ensure(bool condition, string message = null)
+        {
+            if (!condition)
+            {
+                throw new InvalidOperationException(message ?? "Invalid expression.");
+            }
+        }
+
+        private static string GetPropertyName(Expression<Func<T, object>> selector)
+        {
+            LambdaExpression lambda = selector as LambdaExpression;
+            Ensure(lambda != null);
+            Ensure(lambda.Parameters.Count == 1);
+
+            ParameterExpression param = lambda.Parameters[0];
+            Ensure(param.Type == typeof(T));
+
+            Expression body = lambda.Body;
+            UnaryExpression conversion = body as UnaryExpression;
+
+            if (conversion != null)
+            {
+                Ensure(conversion.NodeType == ExpressionType.Convert);
+                body = conversion.Operand;
+            }
+
+            MemberExpression member = body as MemberExpression;
+            Ensure(member != null);
+            Ensure(member.Expression == param);
+            Ensure(member.Member.MemberType == System.Reflection.MemberTypes.Property);
+
+            return member.Member.Name;
+        }
+
+        /// <summary>
+        /// Gets the string representation of the built query.
+        /// </summary>
+        /// <returns>String representation of query.</returns>
+        public string GetQueryText()
+        {
+            if (string.IsNullOrEmpty(_queryText))
```

Comment:
For correctness, we need to clear `_queryText` in any of the methods that mutate state.  Calling `.WhereCustom("...")` then `.GetQueryText()` then `.WhereCustom("...")` again shouldn't return the same result when I call `.GetQueryText()` again.

If we're in a hurry, we should just stop caching and build this every time.